Add 1.0 roadmap blog. (#19)
CHERIoT-Platform/CHERIoT-Platform.github.io: _posts/2024-10-31-isa-roadmap.markdown

It is defining a set of extensions under the Zcheri umbrella.
It also has to produce something that is a minimum viable product, with a focus on memory safety.
A lot of things in CHERIoT have demonstrated value in our use case will not make it into a 1.0 release of the standard CHERI extensions.

The things added to RISC-V by CHERIoT can be split into several categories:

 - Things that we think everyone should do.
 - Things where we've made compromises in a direction are necessary for small devices.
 - Things where we've co-designed the hardware and software and, while we think they're the right approach in a purist sense, may not be the most compatible options.

We've been working with the RISC-V International CHERI SIG and TG on ensuring that the Zcheripurecap extension is a subset of CHERIoT.
We believe that, in the current drafts, everything in our first category is in the RVI standard and everything in the second and third categories are permitted.

This will become important for CHERIoT 2.0.

CHERIoT ISA v 2.0
-----------------

Our goal with the 2.0 release of the CHERIoT ISA is to rebase on top of the official Zcheripurecap extension, once this has been ratified.

> We expect to be able to provide compatibility between 1.0 and 2.0, such that anything that isn't machine code can run on both.

Reviewer: ```suggestion
We expect to be able to provide source compatibility between 1.0 and 2.0, such that anything that isn't machine code can run on both.
```

Author: I didn't want to say 'source' because assembly isn't usually in the space of 'source'.

Reviewer: That's fine; stet.

> RISC-V International will assign new opcodes for the CHERI instructions and so this release won't be binary compatible, but we expect to be able to maintain source compatibility.

Reviewer: ```suggestion
We anticipate that RISC-V International will assign new opcodes for the CHERI instructions and so this release won't be binary compatible, but we expect to be able to maintain source compatibility.
```

> This will likely require a compatibility file of assembly macros because the official standard renames most instructions to make them ~~less readable~~ more consistent with RISC-V naming conventions.

Reviewer: While I agree... perhaps
```suggestion
The official standard also renames most instruction mnemonics to make them more consistent with RISC-V naming conventions.
To further ease source transition, we will provide a compatibility file of assembly macros for the current instruction names.
```

For C and higher-level languages, the compiler will produce different instructions, but with the same abstract machine.

This compatibility means that it should be trivial to support 1.0 and 2.0 for the same amount of time.

We may propose other parts of our extensions as future Zcheri standards, if they have value beyond embedded devices.

What does CHERIoT have that Zcheri doesn't?
-------------------------------------------

CHERIoT makes a lot of use of one of the concepts that's been in CHERI since the beginning: sealing.
Sealing is how we build type-safe opaque types.
This, in turn, is how we build software capability systems.

> I believe sealing is essential to providing a robust compartmentalisation model, but it's not necessary for single-trust-domain memory safety.

Reviewer: ```suggestion
We believe sealing is essential to providing a robust compartmentalisation model, but it's not necessary for single-trust-domain memory safety.
```

We believe that our local/global behaviours can be mapped to the underlying model as a simple two-level case for this.

The biggest place where CHERIoT diverges from RISC-V is in reducing the shift of the `AUIPCC` instruction (the capability version of `AUIPC`).
In RISC-V, materialising a PC-relative address is done in two stages.
First, AUIPC adds an immediate value that is left-shifted by 12 to PC, and then a subsequent instruction (either an add, or the immediate in a load or store) adds in the remaining displacement.

Unfortunately, this means that (in the worst case) the result of the `AUIPC` instruction can go 4095 bytes too far.
On normal RISC-V, that's fine: it's just an integer.
With CHERI, this may be 4095 bytes out of bounds of the program counter capability.

When you scale a capability encoding down to 32 bits and short pipelines, representing 4 KiB out-of-bounds values adds some unfortunate complexity.
We have some proposed formats that will allow this, but microarchitects have estimated that it will hit the maximum frequency of short pipelines by as much as 10%.

> In CHERIoT, we reduce the shift in `AUIPCC` by one.

Reviewer: ```suggestion
To accommodate CHERIoT v1's capability encoding, we reduce the shift in `AUIPCC` by one.
```

The current Zcheri drafts assume the existence of an 'omnipotent' capability.
A capability that allows read, write, and execute is important for 100% backwards compatibility with APIs such as `mmap` (whether this is desirable is an exercise for the reader), but on CHERIoT we do not have such APIs.
CHERIoT has a *multi-rooted* capability hierarchy.
We have a read-execute root, a read-write root, and a sealing root.
These all have different permission encodings and so it is not possible to construct a write-execute capability (though you can have two capabilities to the same memory, one with write and one with execute permissions).

The remaining differences are fairly small.
For example, we have a set-bounds instruction with a large immediate.
This is important because our ABI sets bounds on globals.
This matters less on the larger systems for two reasons.
First, they typically use an ABI where pointers to globals can be loaded via a cap table, so bounds are not typically applied dynamically.

> Second, larger systems have a different distribution of sizes for globals and so the 12-bit immediate is less useful (it hits almost 100% of cases for embedded systems).

Reviewer: ```suggestion
Second, while small systems tend to have correspondingly small global objects, so the 12-bit immediate covers nearly 100% of cases, larger systems have a different and wider distribution of sizes for their globals.
```
